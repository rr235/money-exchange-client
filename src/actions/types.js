export const FETCH_POCKETS = 'fetch_pockets';
export const SELECT_POCKET_FROM = 'select_pocket_from';
export const SELECT_POCKET_TO = 'select_pocket_to';
export const SET_AMOUNT_FROM = 'set_amount_from';
export const SET_AMOUNT_TO = 'set_amount_to';
export const SET_EXCHANGE_RATE = 'set_exchange_rate';
