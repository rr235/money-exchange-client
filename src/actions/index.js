import axios from 'axios';
import {
  FETCH_POCKETS,
  SELECT_POCKET_FROM,
  SELECT_POCKET_TO,
  SET_AMOUNT_FROM,
  SET_AMOUNT_TO,
  SET_EXCHANGE_RATE,
} from './types';

const dispatchExchangeRate = async (dispatch, from, to) => {
  const { data } = await axios.get(
    `http://localhost:5000/exchange?from=${from}&to=${to}`
  );
  dispatch({ type: SET_EXCHANGE_RATE, payload: data.conversionRate });
};

export const fetchPockets = () => async (dispatch) => {
  const { data } = await axios.get('http://localhost:5000/pockets');
  dispatch({
    type: FETCH_POCKETS,
    payload: data,
  });
  dispatch({ type: SELECT_POCKET_FROM, payload: data[0] });
  dispatch({ type: SELECT_POCKET_TO, payload: data[1] });

  // set exchange rate
  dispatchExchangeRate(dispatch, data[0].code, data[2].code);
};

export const selectPocketFrom = (pocket) => (dispatch) => {
  dispatch({ type: SELECT_POCKET_FROM, payload: pocket });
};

export const selectPocketTo = (pocket) => (dispatch) => {
  dispatch({ type: SELECT_POCKET_TO, payload: pocket });
};

export const setAmountFrom = (amount) => (dispatch) => {
  dispatch({ type: SET_AMOUNT_FROM, payload: amount });
};

export const setAmountTo = (amount) => (dispatch) => {
  dispatch({ type: SET_AMOUNT_TO, payload: amount });
};

export const exchangeAmountFrom = ({ amount, rate }) => async (dispatch) => {
  dispatch({
    type: SET_AMOUNT_TO,
    payload: Number((amount * rate).toFixed(2)),
  });
};

export const exchangeAmountTo = ({ amount, rate }) => async (dispatch) => {
  dispatch({
    type: SET_AMOUNT_FROM,
    payload: Number((amount / rate).toFixed(2)),
  });
};

export const setExchangeRate = ({ from, to }) => async (dispatch) =>
  dispatchExchangeRate(dispatch, from, to);
