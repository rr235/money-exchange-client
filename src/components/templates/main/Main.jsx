import React, { Component } from 'react';
import styles from './main.styles.scss';
import Button from '../../atoms/button';
import Input from '../../atoms/input';
import Dropdown from '../../atoms/dropdown';

class Main extends Component {
  state = {
    options: [
      { text: 'foo', value: 1 },
      { text: 'bar', value: 2, selected: true },
      { text: 'baz', value: 3 },
    ],
  };

  getOptions = () => {
    return this.props.pockets.map(({ code }) => ({ text: code, value: code }));
  };

  selectionHandler = ({ value: selectedValue }) => {
    const updatedOptions = this.state.options.map(({ text, value }) => ({
      text,
      value,
      selected: value === selectedValue,
    }));
    this.setState({ options: updatedOptions });
  };

  render() {
    return (
      <div className={styles.main}>
        <Button>Button</Button>
        <Dropdown
          options={this.state.options}
          onSelect={this.selectionHandler}
        />
        <Input label="foo" id="txtFoo" />
      </div>
    );
  }
}

export default Main;
