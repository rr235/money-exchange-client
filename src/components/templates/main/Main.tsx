import React from 'react';
import styles from './main.styles.scss';

const Main = () => <div className={styles.main}>This is main component.</div>;

export default Main;
