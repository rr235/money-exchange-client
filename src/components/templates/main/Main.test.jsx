import React from 'react';
import { mount } from 'enzyme';
import configureStore from 'redux-mock-store';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import Main from './Main';
import { fetchPockets } from '../../../actions';

jest.mock('../../../actions');

const mockStore = configureStore([thunk]);

const storeData = {
  pockets: [
    {
      name: 'Euro',
      code: 'EUR',
      symbol: '€',
      balance: 10,
    },
    {
      name: 'Pound Sterling',
      code: 'GBP',
      symbol: '£',
      balance: 10,
    },
    {
      name: 'US Dollar',
      code: 'USD',
      symbol: '$',
      balance: 10,
    },
  ],
  from: {
    pocket: {
      name: 'Euro',
      code: 'EUR',
      symbol: '€',
      balance: 10,
    },
    amount: 0,
    exceedsBalance: false,
  },
  to: {
    pocket: {
      name: 'Pound Sterling',
      code: 'GBP',
      symbol: '£',
      balance: 10,
    },
    amount: 0,
  },
  exchangeRate: 0.87467,
};

describe('Main', () => {
  let store;
  let component;

  beforeEach(() => {
    store = mockStore(storeData);
    store.dispatch = jest.fn();

    component = mount(
      <Provider store={store}>
        <Main />
      </Provider>
    );
  });

  it('should render correctly', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <Main />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should fetch pockets data', () => {
    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(fetchPockets).toHaveBeenCalled();
  });

  it('from input should have value 0', () => {
    const fromInput = component.find('input[id="input-from"]');
    expect(fromInput.prop('value')).toBe(0);
  });

  it('to input should have value 0', () => {
    const toInput = component.find('input[id="input-to"]');
    expect(toInput.prop('value')).toBe(0);
  });

  it('"from" dropdown should show correct pocket', () => {
    const fromDropdown = component.find('button[id="select-from"]');
    expect(fromDropdown.text()).toBe(storeData.from.pocket.code);
  });

  it('"to" dropdown should show correct pocket', () => {
    const toDropdown = component.find('button[id="select-to"]');
    expect(toDropdown.text()).toBe(storeData.to.pocket.code);
  });
});
