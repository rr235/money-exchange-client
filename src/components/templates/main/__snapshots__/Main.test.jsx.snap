// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main should render correctly 1`] = `
<form
  className="main"
>
  <div
    className="content"
  >
    <span
      className="exchangeRate"
    >
      1 EUR = 0.87467 GBP
    </span>
    <div
      className="selector"
    >
      <div>
        <div
          className="dropdown dropdown"
        >
          <button
            aria-haspopup="listbox"
            className="button"
            id="select-from"
            onClick={[Function]}
            type="button"
          >
            EUR
          </button>
          <ul
            className="options"
            role="listbox"
          >
            <li
              aria-selected={false}
              className="option"
              data-value="EUR"
              onClick={[Function]}
              onKeyDown={[Function]}
              role="option"
              tabIndex="0"
            >
              EUR
            </li>
            <li
              aria-selected={false}
              className="option"
              data-value="GBP"
              onClick={[Function]}
              onKeyDown={[Function]}
              role="option"
              tabIndex="0"
            >
              GBP
            </li>
            <li
              aria-selected={false}
              className="option"
              data-value="USD"
              onClick={[Function]}
              onKeyDown={[Function]}
              role="option"
              tabIndex="0"
            >
              USD
            </li>
          </ul>
        </div>
        <span
          className="balance"
        >
          Balance: 
          €
          10
        </span>
        <label
          className="label"
          htmlFor="input-from"
        >
          From (-)
        </label>
        <input
          autoComplete="off"
          className="input"
          id="input-from"
          onChange={[Function]}
          type="text"
          value={0}
        />
        <div
          className="exceedsBalanceMessage"
        />
      </div>
    </div>
    <div
      className="selector"
    >
      <div>
        <div
          className="dropdown dropdown"
        >
          <button
            aria-haspopup="listbox"
            className="button"
            id="select-to"
            onClick={[Function]}
            type="button"
          >
            GBP
          </button>
          <ul
            className="options"
            role="listbox"
          >
            <li
              aria-selected={false}
              className="option"
              data-value="EUR"
              onClick={[Function]}
              onKeyDown={[Function]}
              role="option"
              tabIndex="0"
            >
              EUR
            </li>
            <li
              aria-selected={false}
              className="option"
              data-value="GBP"
              onClick={[Function]}
              onKeyDown={[Function]}
              role="option"
              tabIndex="0"
            >
              GBP
            </li>
            <li
              aria-selected={false}
              className="option"
              data-value="USD"
              onClick={[Function]}
              onKeyDown={[Function]}
              role="option"
              tabIndex="0"
            >
              USD
            </li>
          </ul>
        </div>
        <span
          className="balance"
        >
          Balance: 
          £
          10
        </span>
        <label
          className="label"
          htmlFor="input-to"
        >
          To (+)
        </label>
        <input
          autoComplete="off"
          className="input"
          id="input-to"
          onChange={[Function]}
          type="text"
          value={0}
        />
        <div
          className="exceedsBalanceMessage"
        />
      </div>
    </div>
    <div
      className="buttonWrapper"
    >
      <button
        className="button button"
        disabled={false}
        id="btnExchange"
        onClick={[Function]}
        type="submit"
      >
        Exchange
      </button>
    </div>
  </div>
</form>
`;
